### ISSUE TITLE
~literal doesn't recognize macros in other modules

### ISSUE THREAD

**Issue by jsmaniac** [racket/typed-racket#329]
### What version of Racket are you using?

Latest nightly: 6.4.0.15 (20160403-794061b)

### What program did you run?

```racket
#lang typed/racket

(module m typed/racket
  (provide foo)
  (require (for-syntax syntax/parse))
  (define-syntax (foo stx)
    (syntax-parse stx
      [(_ (~literal foo)) #''yes]
      [(_ x) #''no]))

  (foo foo)) ;; Gives 'yes, as expected

(require 'm)
(foo foo) ;; Gives 'no
```

### What should have happened?

The second call to `(foo foo)`, in the outer module, should produce `#''yes` just like the call inside the `m` module.

I think that when TR wraps the macros so that they print `Type Checker: Macro xxx from typed module used in untyped code` in untyped contexts this has the unwanted side-effect of exporting the macro as a different identifier.

I'm not sure if this can be solved easily, but if you find a workaround to implement a macro recognizing its identifier in the arguments, I'll add it to the in the "gotchas" section of the documentation and submit a PR. For now I'm just using `(~datum foo)` instead, but it's a bit too permissive, and doesn't handle `(require (rename-in 'm [foo renamedfoo]))` properly.

**Comment by samth** [comment]
I don't think this can be fixed, so just documenting it seems like the right thing to me.

**Comment by jsmaniac** [comment]
@samth I just found this workaround. Renaming and hygiene work as expected with this. Does it seem reasonable enough for me to include it in the docs, or is there some obscure gotcha with this technique?

```
#lang typed/racket

(module m-foo-id typed/racket
  (provide foo)
  (begin-for-syntax
    (provide foo-impl)
    (define foo-impl
      (box (λ _
             (error "Implementation for foo was not loaded!")))))
  (define-syntax (foo stx)
    ((unbox foo-impl) stx)))

(module m typed/racket
  (provide foo)
  (require (submod ".." m-foo-id))
  (require (for-syntax syntax/parse))
  (begin-for-syntax
    (set-box! foo-impl
              (λ (stx)
                (syntax-parse stx
                  [(_ (~literal foo)) #''yes]
                  [(_ x) #''no]))))
  
  (foo foo)) ;; Gives 'yes, as expected

(require 'm)
(foo foo) ;; Gives 'yes, hooray!
```

**Comment by samth** [comment]
Yes, that should work. I think you should be able to make something simpler by providing the identifier as an argument, but I haven't thought it through yet.
